### ISSUE TITLE
Specifying the "select" param in a query using a variable with a defined type breaks types 

### ISSUE THREAD

**Issue by artemzakharov** [prisma/prisma#3372]
## Bug description

After upgrading Prisma from 2.2 to 2.5, I'm now running into an issue where, when I try to provide the **select** parameter using a typed variable rather than inline code  in a query, the result of the query no longer has any type information associated with it. 

For example, I have a User type that has a number of properties, like id, createdAt, etc. 

If I attempt to do a query with the **select** being defined inline, types are present on the result as expected:

![image](https://user-images.githubusercontent.com/25446299/90814949-c4c3ba00-e2de-11ea-9184-9aa38d5bee1d.png)

If I now define my selection property using a typed variable, however (with the type corresponding to to the type expected by the Prisma for the query), and the pass that in instead, the result no longer has any properties, being typed as an empty object instead:

![image](https://user-images.githubusercontent.com/25446299/90815151-19ffcb80-e2df-11ea-89a0-cf24680219de.png)

Oddly, if I remove the typing on the select variable, type information returns to the result (but I no longer have type suggestions when composing the select variable)

![image](https://user-images.githubusercontent.com/25446299/90815631-d2c60a80-e2df-11ea-9fb6-ee204f5d4d65.png)

## How to reproduce

Take any query, extract the **select** property into a typed variable.

## Expected behavior

I expect the result of the query to have type information that corresponds to what the select variable is set to.

## Prisma information

This appears to apply to any and all schemas, types, and queries. 

## Environment & setup

- OS: Windows
- Database: PostgreSQL
- Node.js version: 10.16
- Prisma version: 2.5.1

**Comment by vh13294** [comment]
related issue https://github.com/prisma/prisma-client-js/issues/810

**Comment by pantharshit00** [comment]
Hey @vh13294 and @artemzakharov 

Sorry I didn't explain this problem properly in https://github.com/prisma/prisma-client-js/issues/810#issuecomment-670635355, apologies. Let me explain the problem here and suggest you the problem solution.

The problem here is that when you type `userSelection` with `UserSelect`, it can't really convey the truthy keys as you have widen the type of that object to UserSelect.

Now when we try to narrow the type down of the User using this mapped type internally `[P in TrueKeys<S['select']>]: P extends keyof User ? User[P]` , we can't find any truthy keys in the `type S['select']` so at the end it just returns `{}` and causes you type issues.

In order to solve this, you will need to narrow the type using a function which takes a generic argument like so:

```ts
type CheckSelectKeys<T, U> = {
  [K in keyof T]: K extends keyof U ? T[K] : never;
};

const createUserSelect = <T extends UserSelect>(
  arg: CheckSelectKeys<T, UserSelect>
) => arg;

const testSelect = createUserSelect({
  name: true,
  email: true,
});
```

The above code will narrow the type of testSelect and it will also check if it doesn't contain any invalid key. 

![image](https://user-images.githubusercontent.com/22195362/92033161-d8642b80-ed88-11ea-8c6a-68c605907950.png)


![image](https://user-images.githubusercontent.com/22195362/92033214-f3cf3680-ed88-11ea-976a-30cf381fc40d.png)

Then you can use it in your code.

**Comment by vh13294** [comment]
@pantharshit00 

Can I just import Subset instead of creating CheckSelectKeys?

```typescript
export declare type Subset<T, U> = {
  [key in keyof T]: key extends keyof U ? T[key] : never;
};

import { Subset } from "@prisma/client";
```

**Comment by pantharshit00** [comment]
Oh yeah, I actually forgot that we export that.

**Comment by timsuchanek** [comment]
Thanks @artemzakharov for reporting. As @pantharshit00 correctly noted, this is intended to not work anymore.

The reason is, that if you typecast it to the `Photon.UserSelect` type, that this is not type safe anymore.
Simple example:

```ts
type UserSelect {
  id: boolean
  name: boolean
}

const select: UserSelect = { id: true }

const result = await prisma.user.findMany({ select })
```

In this example `result` will now also include `name` in the type, as the **type** of `select` is `UserSelect` which also has the `name`, even though `select` doesn't have.
In other words, Prisma Client is not looking into the values but just the types for the ts type validation / inference.

Exactly for this situation that you described we introduced more strict type checks, that make sure, that you don't just import `UserSelect` from `@prisma/client` and type cast the client, as that would break the type safety.

The only solution we have within TypeScript for this, is, to remove the type cast of your `userSelection` variable and let TypeScript validate things later in the `findOne` call.
I know that's not the 100% optimal solution, but for now, this is all we can do here.

Something we could definitely improve - we should in the types detect this case and instead of having the empty type actually generate a nice type-level error message, that describes the problem and the solution.

**Comment by MichalLytek** [comment]
@timsuchanek 
What is the recommended way to create business logic wrappers around Prisma Client?
Translating function params to prisma `where` object is easy but I can't simply make the returned type typesafe while allowing consumers to request only selected fields (or relations):

```ts
function findNewPrismaUsersInNearby(select: Prisma.UserSelect) {
  return prisma.user.findMany({
    where: {
      firstName: {
        contains: 'Prisma',
      },
      country: {
        in: ['Germany', 'Poland'],
      },
      createdAt: {
        gte: new Date('2020-12-20'),
      },
    },
    select,
  });
}
```
This narrows the returned object type to `{}`:
```ts
function findNewPrismaUsersInNearby(select: Prisma.UserSelect): Promise<{}[]>
```

**Comment by Sytten** [comment]
Just a suggestion out of the blue. Maybe if the input type is a generic that extends the select type it would work. Basically it would refine the typing of `field: boolean` to `field: true` which in turn should allow prisma to build the output type. I have not tried though so not sure if it works.

**Comment by timsuchanek** [comment]
Thanks for the answer @Sytten! Yes I totally think we can add something there.
I'll candidate this for next sprint.

**Comment by timsuchanek** [comment]
Thanks for all the comments.
We are talking about two distinct problems here.

#### 1. Problem: How do I separately construct the `select` or `include` argument and not just inline?
#### 2. Problem: How do I create a wrapper function around any Prisma query, like `findMany` that is typesafe?

These problems require separate solutions. In the following, I'll show how they're solvable today and also mention a potential utility type that we might want to generate in the client to save you some typing.

### 1. Problem: Dynamically construct `select` or `include`
This is, what @artemzakharov and @vh13294 requested.
This can be achieved with a little helper function, as Harshit already noted. Here called `makeUserSelect`:

```ts
import { PrismaClient, Prisma } from './@prisma/client'

function makeUserSelect<T extends Prisma.UserSelect>(
  select: Prisma.Subset<T, Prisma.UserSelect>,
): T {
  return select
}

const select = makeUserSelect({ id: true })

const user = await prisma.user.findFirst({
  where: {
    id: 'id',
  },
  select,
})

const id = user?.id
```
The same for `include`:

```ts
function makeUserInclude<T extends Prisma.UserInclude>(
  include: Prisma.Subset<T, Prisma.UserInclude>,
): T {
  return include
}

const include = makeUserInclude({
  likes: true
})

const user = await prisma.user.findFirst({
  where: {
    id: 'id',
  },
  include,
})

const likes = user?.likes
```
We can even do this for the whole args, here an example for `findMany`:

```ts
function makeUserFindManyArgs<T extends Prisma.UserFindManyArgs>(
  include: Prisma.Subset<T, Prisma.UserFindManyArgs>,
): T {
  return include
}

const args = makeUserFindManyArgs({
  include: {
    likes: true
  }
})

const users = await prisma.user.findMany(args)

const likes = users[0].likes
```

With our current knowledge, we need to create one function per type we want to "make". We will later investigate if we could simplify that.

### 2. Problem: How do I create a wrapper function around any Prisma query, like `findMany` that is typesafe?
This is, what @MichalLytek and @Sytten talk about.
This can be done without any helper type.
We can make this happen both for `select` and `include`, but not the whole args:

Just passing in `select`, as requested by @MichalLytek 
```ts
function findNewPrismaUsersInNearbySelect<S extends Prisma.UserSelect>(
  select: Prisma.Subset<S, Prisma.UserSelect>,
) {
  return prisma.user.findMany<{ select: S }>({ select, where: {} })
}

const users = await findNewPrismaUsersInNerbySelect({ id: true })
users[0].id
```
The same for `include`:
```ts
function findNewPrismaUsersInNearbyInclude<S extends Prisma.UserInclude>(
  include: Prisma.Subset<S, Prisma.UserInclude>,
) {
  return prisma.user.findMany<{ include: S }>({ include })
}
const users = await findNewPrismaUsersInNearbyInclude({ posts: true })
users[0].posts
```
**We can't** do this for the whole args here:
```ts
function findNewPrismaUsersInNearby<T extends Prisma.UserFindManyArgs>(
  args: Prisma.Subset<T, Prisma.UserFindManyArgs>,
) {
  return prisma.user.findMany({...args, where: {}})
}

const users = await findNewPrismaUsersInNearby({ select: { id: true } })
// types will be incorrect
```
But that also hasn't been asked 🙂 

**Does this solve your problem? Please let us know 🙏** 

Next steps for us at Prisma:
- [ ] Document this
- [ ] Find out, which `make...` types we want to expose

**Comment by timsuchanek** [comment]
We just got help from a TypeScript god - Pierre-Antoine Mills @millsp, the author of ts-toolbelt (without his work, we wouldn’t have group by today).
He has some ideas: https://github.com/timsuchanek/make-type/blob/master/draft.ts
Unfortunately, TypeScript has an annoying limitation, that you can’t “skip” type parameters.
Therefore, it looks like we have to generate the `createUserSelect` and `createUserInclude` API for all types.

**Comment by timsuchanek** [comment]
What do you think about this:
```ts
const userValidator = validator<Prisma.UserSelect>();

/**
 * Otherwise, since the .find methods also perform type checks, then why not
 * just use `Narrow` and let it do its job. This is a more elegant solution:
 */
declare function select<S>(select: Narrow<S>): S;

async function main() {
  const prisma = new PrismaClient()

  const select0 = userValidator({ id: true, name: true });  // narrowed

  prisma.$disconnect();
}
```

I quite like this! This would save us generating a lot of types, not making the generated client so bloated.

**Comment by MichalLytek** [comment]
> Unfortunately, TypeScript has an annoying limitation, that you can’t “skip” type parameters.

I always use this trick:
```ts
function wrapper<TExplicitType>() {
  return function <TImplictType>(t: TImplictType): TImplictType & TExplicitType {
    // ...
  }
}

const foobar = wrapper<Foo>()(bar);
```

**Comment by millsp** [comment]
Definitely, that's the workaround the validator uses:

```ts
declare function validator<V>(): <S>(select: Exact<Narrow<S, V>, V>) => S;
```

I'm working on a codebase that relies (too) much on this - it's heavy for my cognitive load. 
But there is a potential PR candidate that should solve this problem in the near (?) future:
https://github.com/microsoft/TypeScript/pull/26349

The TypeScript team has added this on their roadmap, but no date is yet set. 
This is one of the most exciting (missing) features in TypeScript today, IMO.

**Comment by MichalLytek** [comment]
> `return prisma.user.findMany<{ select: S }>({ select, where: {} })`

@timsuchanek this snippets gives a TS error 😛 

```
Argument of type '{ select: Prisma.Subset<S, Prisma.UserSelect>; where: {}; }' is not assignable to parameter of type '{ select: S; }'.
  Object literal may only specify known properties, and 'where' does not exist in type '{ select: S; }'.
```

It has to be explicitly `.findMany<{ select: S, where: Prisma.UserWhereInput}>(...)` which adds more boilerplate if you want to add more params like `rejectOnNotFound: true` 😞 

Second case, I can't make this approach work if I want the `select` to be optional like with direct prisma client usage. If I make this parameter optional, the return type becomes `{}` 😕

**Comment by millsp** [comment]
Errors did not show up when *only* bad fields were passed:
```ts
  userValidator({ asd: true });
```

This came from a mistake I made when writing `Exact`. I was trying to improve intellisense errors by doing a type cast... which caused an untested case to fail. So I rewrote it, and all errors are caught properly (with nice intellisense errors):

```ts
type Exact<A, W> = {
    [K in keyof A]: K extends keyof W
    ? A[K] extends W[K] ? A[K] : W[K]
    : never
}
```

`Exact` takes any object `A` and makes it comply by `W`. If `A` has prop that is in not in `W`, we mark it as never. If `A[K]` complies by `W[K]` we return `A[K]`, otherwise we return `W[K]` and this will give us better intellisense errors in case of conflicts.

**Comment by millsp** [comment]
Errors did not show up when *only* bad fields were passed:
```ts
  userValidator({ asd: true });
```

This came from a mistake I made when writing `Exact`. It is a type utility that forces a type to comply by another one - with not more and not less properties. I was **trying to improve intellisense errors** by doing a **type cast**... which caused an untested case to fail. So I rewrote it, and all errors are caught properly (with nice intellisense errors):

```ts
type Exact<A, W> = 
A extends W ? W extends unknown ? 
A extends NotMappableType ? A : {
    [K in keyof A]: K extends keyof W
    ? Exact<A[K], W[K]>
    : never
} : never : W

type NotMappableType =
| string
| number
| bigint
| boolean
| ((...args: any) => any);
```

`Exact` takes any object `A` and makes it comply by `W`. If `A` has a prop that is in not in `W`, we mark it as never. If `A[K]` complies by `W[K]` we return `A[K]`, otherwise we return `W[K]` and this will give us better intellisense errors in case of conflicts.

This is a more mature `Exact` type utility than the ones proposed in https://github.com/microsoft/TypeScript/issues/12936#issuecomment-368244671 since it is able to:
(community-developed `Exact` utilities all fail at least one of the points below (where points 1 and 2 are the most important IMO).
1. Give explicit errors about types at any depth
2. Does not fail on mixed unions/distributes well
3. Narrow the user input given that it is correct

(If you're interested to see the comparison https://github.com/millsp/make-type/blob/06d9cbd63100ad27b9b950998a34d1d48627bafc/draft.ts#L42-L79)

**Comment by timsuchanek** [comment]
> @timsuchanek this snippets gives a TS error 😛

Oops, I think I only tested it without the `where` on my machine 🙈 
I see this as a possible solution to make it more general:
```ts
function findNewPrismaUsersInNearbySelect<S extends Prisma.UserSelect>(
  select: Prisma.Subset<S, Prisma.UserSelect>,
) {
  return prisma.user.findFirst<
    { select: S } & Omit<Prisma.UserFindFirstArgs, 'select' | 'include'>
  >({ select })
}
```
@millsp do you maybe have an idea how to make that more elegant?
You can use https://github.com/timsuchanek/make-type/blob/master/main.ts as a playground

**Comment by timsuchanek** [comment]
And @millsp thanks a lot for the help there! I just merged your new types and they work like a charm!

**Comment by dilizarov** [comment]
@timsuchanek Thanks for the code snippet above. It has helped be wrestle with an issue I had... I'm not sure I understand this piece though: `{ select: S } & Omit<Prisma.UserFindFirstArgs, 'select' | 'include'>`

Why is the `Omit` block necessary?

**Comment by pantharshit00** [comment]
@dilizarov Omit block is not necessary. It is there so that you can use other keys like `where` in that query in future and the parameter which the function take is properly typed for that.

**Comment by pip8786** [comment]
@MichalLytek I can't quite get it compiling with your solution. How does the return type of the last function encompass when the select parameter is omitted? Are all 3 function definitions needed? Would you have an example where the { ... } is fully expanded just to make sure I understand how the parameters are used?

**Comment by Cretezy** [comment]
The following gave me an error:
```ts
  getUserById = async <T extends Prisma.UserInclude>(
    id: string,
    include: T | null = null,
  ): Promise<Prisma.UserGetPayload<{ include: T }> | null> => {
    // Type error here, for return value type
    return await this.prisma.user.findUnique({
      where: {
        id,
      },
      include,
    });
  };
```

The following worked (with `Prisma.Subset` in the return type):
```ts
  getUserById = async <T extends Prisma.UserInclude>(
    id: string,
    include: T | null = null,
  ): Promise<Prisma.UserGetPayload<{ include: Prisma.Subset<Prisma.UserInclude, T> }> | null> => {
    return this.prisma.user.findUnique({
      where: {
        id,
      },
      include,
    });
  };
```

**Comment by stephyswe** [comment]
so, essentially, we trying to create a flexible function. where include can be anything ?
Also how would you use this function? 

#Error - not finding include - driver.
![image](https://user-images.githubusercontent.com/20955623/181694836-747bd411-d9d4-4dc5-be24-2082a2cb05bf.png)


Attempt #1 - with order
```js
const order = await this.findOrderById(id, { driver: true });
      if (order.driver.id) {
        return {
          ok: false,
          error: 'This order already has a driver',
        };
      }
      
findOrderById = async <T extends Prisma.OrderInclude>(
    id: number,
    include: T | null = null,
  ): Promise<Prisma.OrderGetPayload<{
    include: Prisma.Subset<Prisma.OrderInclude, T>;
  }> | null> => {
    return this.prisma.order.findUnique({
      where: {
        id,
      },
      include,
    });
  };
  ```

**Comment by andrewmclagan** [comment]
Why is this issue closed? Its so hard to work with Prisma when you are always fighting against its type system with unpredictable results.

**Comment by janpio** [comment]
@andrewmclagan https://github.com/prisma/prisma/issues/3372#issuecomment-773196159 If there still is a problem, please open a new issue.

**Comment by Azoom-Le-Van-Thuong** [comment]
@stephyswe  I should try code below: 
```
  async getListUserBranch<T extends Prisma.UserBranchInclude>(
    whereClause: Prisma.UserBranchWhereInput,
    include: T | null = null,
  ): Promise<
    Prisma.UserBranchGetPayload<{
      include: T
    }>[]
  > {
    return await this.prismaService.userBranch.findMany({
      where: whereClause,
      include,
    })
  }
```

**Comment by emazzu** [comment]
Hi, 
Loopback framework, permit pass where, order, etc, by querystring natively, is it possible in prisma ????

http://[domainName]:3000/api/goNotes?filter[include]=productionMethod&filter[include]=installation&filter[include]=actionType&filter[include]=action&filter[include]=actionDetail&filter[include]=revisionType&filter[include]=resource&filter[where][created][between][0]=2023-02-06&filter[where][created][between][1]=2023-02-16:23:59&filter[order]=created%20DESC

**Comment by TiE23** [comment]
> @MichalLytek I can't quite get it compiling with your solution. How does the return type of the last function encompass when the select parameter is omitted? Are all 3 function definitions needed? Would you have an example where the { ... } is fully expanded just to make sure I understand how the parameters are used?

He's using "Function Overload", and they're really quite something. Read about them, here: https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads

The thing they provide is the ability to differentiate between returned values depending on the arguments provided. So, when calling a function with different arguments, it'll know to anticipate different returned values.

----
Here's my crack at it. None of these interesting `Subset` solutions worked for me. I, too, had to use function overloads.

For example, my old `getCohort` function looked like this:

```typescript
async function getCohort(cohortId: number, relatedUserId?: number) {
  return prisma.cohort.findFirst({
    where: { id: cohortId, ...buildUserRelationCohortFilter(relatedUserId) },
    include: cohortWithBaseRelations.include,
  });
}
```

(The "buildUserRelationCohortFilter()" function is unimportant for this example, but I included it to show my real-world situation).

As you can see, I had one issue to deal with immediately: My function already had an optional argument.

Then, atop of this, I wanted to make sure that `customInclude` had to be completely optional. When I attempted to pass empty objects, I had terrible errors only when I tried to execute the code. The Prisma engine (or whatever it could be called) was NOT a fan of an empty `include`.

So, this was the solution I settled on:

```typescript
type GetCohortArgs = {
  cohortId: number;
  relatedUserId?: number;
};

async function getCohort(args: GetCohortArgs): Promise<Cohort | null>;
async function getCohort<I extends Prisma.CohortInclude>(
  args: GetCohortArgs,
  customInclude: I
): Promise<Prisma.CohortGetPayload<{ include: I }> | null>;
async function getCohort<I extends Prisma.CohortInclude>(
  args: GetCohortArgs,
  customInclude?: I
): Promise<Cohort | Prisma.CohortGetPayload<{ include: I }> | null> {
  const { cohortId, relatedUserId } = args;

  return prisma.cohort.findFirst({
    where: { id: cohortId, ...buildUserRelationCohortFilter(relatedUserId) },
    include: customInclude,
  });
}
```

- _Without_ `customInclude` it returns...
  - `Promise<Cohort | null>`
- _With_ `customInclude` it returns...
  - `Promise<Prisma.CohortGetPayload<{ include: I }> | null>`
- _With_ `customInclude?` it returns...
  - `Promise<Cohort | Prisma.CohortGetPayload<{ include: I }> | null>`

Basically, you're offerering these three signatures. When you're calling the function, you're only choosing between the first two.

The third signature is what you implement the function to. And since `customInclude` is optional in this third signature, you write the function body to this condition (which for me is very easy, but in a past project I had to use an if statement and decide between two returns).

**Comment by kostas-pblworks** [comment]
Thank you for everyone for the helpful responses.

I'm trying to do something similar, a `findById` wrapper, that has already been answered here but with optional paramaters  like so:

```TS
async findById<S extends Prisma.ProjectSelect>(
    projectId: number,
    options?:{ fields?: S}
  ): Promise<Prisma.ProjectGetPayload<{
    select?: S
  }> | null> {
    return this.prismaProject.findFirst
    ({
      where: { id: projectId, deletedAt: null },
      select: options?.fields,
    })
  }
```

While the above compiles, the return type has all fields and not just the one selected. Any ideas maybe? 

Thank you

**Comment by sumann7916** [comment]
```
type SelectArgs = Prisma.BlogSelect;

async function GetAllBlogsByUserId(
  userId : string
): Promise<Blog[]>;

async function GetAllBlogsByUserId<S extends SelectArgs>(
  useId: string,
  args: { select?: S; options?: PageOptions }
): Promise<Prisma.BlogGetPayload<{ select: S }>[]>;

async function GetAllBlogsByUserId<S extends SelectArgs>(
  userId: string,
  args?: { select?: S; options?: PageOptions }
): Promise<
  Blog[] | Prisma.BlogGetPayload<{ select: S }>[]
> {
  return db.ablog.findMany({
    where: {},
    select: args?.select,
    skip: args?.options?.skip,
    take: args?.options?.take,
    orderBy: args?.options?.orderBy
  });
}
```
I am able to get conditonal return type this way. But the problem is when I do 
```
const blogs = await GetAllBlogsByUserId(userId, {options: {}});
```
I get all relational type as well as my return type.

**Comment by francescotau** [comment]
Why does this work?

```
  static async getDietActive<T extends Prisma.DietPrismaInclude>(
    userId: string,
    include: Prisma.Subset<T, Prisma.DietPrismaInclude>,
  ) {
    return await prismaClient.dietPrisma.findFirst({
      where: { userId, active: true },
      include,
    });
  }
```

but this doesn't?

```
  static async getDietActive<T extends Prisma.DietPrismaInclude>(
    userId: string,
    include?: Prisma.Subset<T, Prisma.DietPrismaInclude>,
  ) {
    return await prismaClient.dietPrisma.findFirst({
      where: { userId, active: true },
      include,
    });
  }
```

How can I achieve an optional include parameter?

**Comment by giladv** [comment]
So, my solution - 
```
export async function getSomething<S extends Prisma.AccountSelect | null>(
    select: S,
    otherParam: {...}
  ) {
    return await prisma.account.findFirstOrThrow<{
      select: S;
      where: { id: string };
    }>({
      where: {
        id: 'abc',
      },
      select,
    });
  }
}

```

**Comment by dunak-debug** [comment]
Any news? It's impossible with prisma to pass select with default fields like in this example:

```ts
async validateSession<TSelect extends Prisma.SessionSelect>(
    token: string,
    select: Prisma.Subset<TSelect, Prisma.SessionSelect>,
  ) {
    const tokenHash = await createHash(token);

    const session = await this.prisma.session.findUnique({
      where: { tokenHash },
      // Type 'Subset<TSelect, SessionSelect<DefaultArgs>> & { expiresAt: true; }' is not assignable to type '{ expiresAt: true; }'. Property 'id' does not exist on type '{ expiresAt: true; }'
      select: {
        ...select,
        expiresAt: true,
      },
    });

    if (!session) throw new SessionNotFoundException();
    // Property 'expiresAt' does not exist on type '{ [K in "expiresAt" | keyof TSelect as (Subset<TSelect, SessionSelect<DefaultArgs>> & { expiresAt: true; } & object)[K] extends false...
    if (session.expiresAt < new Date()) throw new SessionExpiredException();

    return session;
  }
```

inside it show ts errors but when calling func type resolves correctly
```ts
const session = await this.sessionService.validateSession(token, {
      ...AUTH_SESSION_SELECT,
      user: { select: AUTH_USER_SELECT },
    });

/*
const session: {
    id: string;
    expiresAt: Date;
    user: {
        id: string;
        role: Role;
    };
}
*/
```
